### PR TITLE
Fixed FP8 fake quantization to use fp32 amax scaling; Added support for FP8 per-channel quantization

### PR DESCRIPTION
**Pull request by realAsma**
## What does this PR do?

**Type of change:** ? new_feature & bug fix

**Overview:** 
For FP8 fake quantization, the QDQ scaling was applied  outside the cuda the kernel in the same precision as the input. This means for bf16 input, the scale is applied in bf16 type. This could cause fake and real quantization to diverge.
Hence in this PR:
1. Fuse QDQ scaling with the QDQ cuda function
2. Remove the previous `fused_fp8` - this only supported per-channel quantization along -1 axis. Per-channel quantization is usually done along 0 axis and this kernel was not useful.
3. Implemented an FP8 per-channel fake quantization kernel (similar to integer per-channel fake quantization)
4. Removed the unnecessary operations such as masking values close to 0 - these add unnecessary overhead and are not present in FP4 or int quantization implementations.
6. Added unit tests.

What was the impact of scaling in bf16 for QAT? 
I have not tested this. I tested the abs quantization error Llama 3.1 8B's first q_proj weight.

Here it is:
| Metric | Value |
|--------|-------|
| Max    | 0.0547 |
| Mean   | 1e-04 |
| Min    | 0.0 |
| Std    | 4e-04 |
| Median | 1.5e-05 |

## Usage
High level usage does not change. See the updated unit tests.

## Testing
Updated 

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**:  the `fused` fp8 fake quantization is removed. However this is not a high-level API and should not cause any impact.
- **Did you write any new necessary tests?**: Yes
- **Did you add or update any necessary documentation?**: NA
- **Did you update [Changelog](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CHANGELOG.rst)?**: No

## Summary by CodeRabbit

- New Features
  - FP8 (E4M3) quantization: per-channel (axis-aware) scaling, optional amax parameter, automatic eager vs GPU-backed selection, outputs preserve original dtype.

- Refactor
  - CPU compiled FP8 implementation removed in favor of a pure-Python eager fallback; CUDA path streamlined, kernels now respect CUDA stream handling; FP8 extension builds GPU-only sources.

- Tests
  - Tests updated for runtime FP8 behavior, expanded per-axis and dtype coverage; legacy test scaffolding removed.

### CODE REVIEW

File: tests/gpu/torch/quantization/test_tensor_quant_cuda.py [hunk 1]
```diff
@@ -187,20 +187,14 @@ def test_non_current_gpu(self, need_2_gpus):
...
-        amax = torch.ones(x.shape[0], 1).cuda() * x.abs().amax()
-        e4m3_x = tensor_quant.scaled_e4m3(x, amax, None, 4, 3)
-        assert torch.allclose(e4m3_x, xq_ref, atol=1e-4, rtol=1e-4)
+    @pytest.mark.parametrize("axis", [0, 1, 2])
+    def test_e4m3_per_channel(self, axis):
+        x = torch.randn(4, 4, 4, dtype=torch.float32).cuda()
+        amax = x.abs().amax(dim=[ax for ax in range(x.ndim) if ax != axis], keepdim=True)
+        scale = 448.0 / amax
+        xq_ref = tensor_quant.scaled_e4m3(x * scale, None, None, 4, 3) / scale
+        xq_test = tensor_quant.scaled_e4m3(x, amax.float(), None, 4, 3)
```
Comment (cjluo-nv) [inline tests/gpu/torch/quantization/test_tensor_quant_cuda.py]: Could you use pure torch impl as the reference?

File: modelopt/torch/quantization/src/tensor_quant_gpu_fp8.cu [hunk 2]
```diff
@@ -31,92 +31,65 @@
...
-  if (amax.numel() == 1) {
-    // case 1.
-    per_block_scaling_factor = false;
-  } else if (amax.numel() > 1 && (amax_ndim > 1 && (amax.size(-1) == amax.numel()))) {
-    // case 2.
-    per_block_scaling_factor = true;
-    blocksize = numel / amax.numel();
-  } else {
-    throw std::runtime_error("invalid combination of inputs and amax shapes/sizes");
-  }
```
Comment (meenchen) [inline modelopt/torch/quantization/src/tensor_quant_gpu_fp8.cu:101]: Are we removing blockwise quantization support from the kernel?

File: tests/gpu/torch/quantization/test_tensor_quant_cuda.py (hunk 1 above)
Comment (realAsma) [reply tests/gpu/torch/quantization/test_tensor_quant_cuda.py]: done.

File: modelopt/torch/quantization/src/tensor_quant_gpu_fp8.cu (hunk 2 above)
Comment (realAsma) [reply modelopt/torch/quantization/src/tensor_quant_gpu_fp8.cu:101]: no. The block wise support did not exist for FP8. As you can see in the removed lines, we are simply raising an error if block wise quantization is detected.

File: modelopt/torch/quantization/src/tensor_quant_gpu_fp8.cu [hunk 3]
```diff
@@ -31,92 +32,70 @@
...
+    fake_e4m3fy_with_axis_cuda_kernel<<<numel / (BLOCK_SIZE * 4) + 1, BLOCK_SIZE>>>(
+        inputs.data_ptr<scalar_t>(), numel, scale.data_ptr<float>(), inv_scale.data_ptr<float>(),
+        axis_size, outer_size, outputs.data_ptr<scalar_t>());
   });
+
   return outputs;
 }
 
-at::Tensor fake_e4m3fy_cuda(at::Tensor inputs) {
+at::Tensor fake_e4m3fy_cuda(at::Tensor inputs, std::optional<at::Tensor> amax_opt) {
```
[the remark this answers is not in the document]
Comment (realAsma) [reply modelopt/torch/quantization/src/tensor_quant_gpu_fp8.cu]: this is fine. `scaled_e4m3_impl` dispatcher checks for this.

File: modelopt/torch/quantization/src/tensor_quant_gpu_fp8.cu [hunk 4]
```diff
@@ -31,92 +32,70 @@
...
-    // case 2.
-    per_block_scaling_factor = true;
-    blocksize = numel / amax.numel();
-  } else {
-    throw std::runtime_error("invalid combination of inputs and amax shapes/sizes");
-  }
-
-  auto stream = c10::cuda::getCurrentCUDAStream();
+  auto scale = 448.f / amax;
+  auto inv_scale = 1.f / scale;
```
Comment (realAsma) [inline modelopt/torch/quantization/src/tensor_quant_gpu_fp8.cu:76]: this is fine. amax should not be zero. we want this case to throw an error - not handle it explicitly.

File: modelopt/torch/quantization/src/tensor_quant_gpu_fp8.cu [hunk 5]
```diff
@@ -31,92 +32,70 @@
...
   }
 }
 
-at::Tensor fused_fake_e4m3fy_cuda(at::Tensor inputs, at::Tensor amax, const float zero_threshold) {
-  size_t numel = inputs.numel();
+at::Tensor fake_e4m3fy_cuda_with_axis(at::Tensor inputs, at::Tensor amax, int axis) {
   auto outputs = torch::empty_like(inputs);
+  size_t numel = inputs.numel();
+  int axis_size = inputs.size(axis);
+  int outer_size = inputs.stride(axis);
```
[the remark this answers is not in the document]
Comment (realAsma) [reply modelopt/torch/quantization/src/tensor_quant_gpu_fp8.cu]: this is fine. contiguous() is checked in the c++ dispatcher `fake_e4m3fy_with_axis`

File: modelopt/torch/quantization/src/tensor_quant_fp8.cpp [hunk 6]
```diff
@@ -18,31 +18,43 @@
...
-at::Tensor fake_e4m3fy(at::Tensor inputs) {
+at::Tensor fake_e4m3fy(at::Tensor inputs, std::optional<at::Tensor> amax) {
+  inputs = inputs.contiguous();
+  if (amax.has_value()) {
+    amax = amax.value().view(-1).to(at::kFloat);
+  }
   if (inputs.is_cuda()) {
-    return fake_e4m3fy_cuda(inputs.contiguous());
+    return fake_e4m3fy_cuda(inputs, amax);
   } else {
```
[the remark this answers is not in the document]
Comment (realAsma) [reply modelopt/torch/quantization/src/tensor_quant_fp8.cpp]: this is fine - lets not handle this.

File: modelopt/torch/quantization/src/tensor_quant_fp8.cpp [hunk 7]
```diff
@@ -18,31 +18,43 @@
...
     return out;
   }
 }
 
-at::Tensor fused_fake_e4m3fy(at::Tensor inputs, at::Tensor amax, const float zero_threshold) {
-  return fused_fake_e4m3fy_cuda(inputs.contiguous(), amax, zero_threshold);
+at::Tensor fake_e4m3fy_with_axis(at::Tensor inputs, at::Tensor amax, int axis) {
+  TORCH_CHECK(inputs.is_cuda());
+  return fake_e4m3fy_cuda_with_axis(inputs.contiguous(), amax.contiguous().to(at::kFloat), axis);
 }
```
[the remark this answers is not in the document]
Comment (realAsma) [reply modelopt/torch/quantization/src/tensor_quant_fp8.cpp]: no this is fine. lets not handles this

File: modelopt/torch/quantization/src/tensor_quant_gpu_fp8.cu [hunk 8]
```diff
@@ -31,92 +32,70 @@
...
 
-  AT_DISPATCH_FLOATING_TYPES(inputs.type().scalarType(), "fused_fake_e4m3fy_cuda", [&] {
-    fused_fake_e4m3fy_kernel<<<numel / (BLOCK_SIZE * 4) + 1, BLOCK_SIZE, 0, stream>>>(
-        inputs.data_ptr<scalar_t>(), numel, amax.data_ptr<float>(), per_block_scaling_factor,
-        blocksize, zero_threshold, outputs.data_ptr<scalar_t>());
+  AT_DISPATCH_FLOATING_TYPES(inputs.type().scalarType(), "fake_e4m3fy_with_axis_cuda", [&] {
+    fake_e4m3fy_with_axis_cuda_kernel<<<numel / (BLOCK_SIZE * 4) + 1, BLOCK_SIZE>>>(
+        inputs.data_ptr<scalar_t>(), numel, scale.data_ptr<float>(), inv_scale.data_ptr<float>(),
+        axis_size, outer_size, outputs.data_ptr<scalar_t>());
   });
```
[the remark this answers is not in the document]
Comment (realAsma) [reply modelopt/torch/quantization/src/tensor_quant_gpu_fp8.cu]: fixed the stream. grid dispatch looks fine to me since it is checked inside the kernel.

File: modelopt/torch/quantization/tensor_quant.py [hunk 9]
```diff
@@ -58,41 +74,16 @@ def scaled_e4m3_impl(
...
-                outputs = cuda_ext_fp8.fake_e4m3fy(inputs * scale) / scale
-
-            # Zero out values that are tiny.
-            # Tiny values could lead to tiny amax and then large scale which cause overflow/saturation
-            # and won't go back to normal value after dividing by scale. The right behavior is to mark them
-            # as zero which also get rid of inf/nan
-            outputs[zero_mask] = 0.0
-
+            outputs = fp8_eager(inputs, amax)
         return outputs
```
[the remark this answers is not in the document]
Comment (realAsma) [reply modelopt/torch/quantization/tensor_quant.py]: not the current method is fine

File: modelopt/torch/quantization/src/tensor_quant_gpu_fp8.cu (hunk 2 above)
Comment (realAsma) [reply modelopt/torch/quantization/src/tensor_quant_gpu_fp8.cu:101]: @meenchen Could you please give another pass? for block quantization, FP8 will fallback to eager mode.

File: modelopt/torch/quantization/tensor_quant.py [hunk 10]
```diff
@@ -42,10 +42,34 @@
...
+    dtype = x.dtype
+    if amax is not None:
+        scale = 448.0 / (amax.to(torch.float32))
+        scale_inv = 1 / scale
+        x = x.to(torch.float32) * scale
+    x = x.to(torch.float8_e4m3fn)
+    if amax is not None:
+        x = x.to(torch.float32) * scale_inv
+    return x.to(dtype)
+
```
[the remark this answers is not in the document]
Comment (realAsma) [reply modelopt/torch/quantization/tensor_quant.py:55]: this is fine.

File: modelopt/torch/quantization/tensor_quant.py [hunk 11]
```diff
@@ -56,43 +80,23 @@ def scaled_e4m3_impl(
...
-            # must be cuda
-            all_cuda = inputs.is_cuda and amax.is_cuda
+    if inputs.is_cpu:
+        return fp8_eager(inputs, amax)
 
-            # also check explicit disable.
-            return amax_last_dim_only and all_cuda and (not disable_fused_kernel)
+    cuda_ext_fp8 = get_cuda_ext_fp8(raise_if_failed=False)
+    if cuda_ext_fp8 is None or amax is None:
+        return fp8_eager(inputs, amax)
```
[the remark this answers is not in the document]
Comment (realAsma) [reply modelopt/torch/quantization/tensor_quant.py:81]: this is fine.

File: modelopt/torch/quantization/tensor_quant.py [hunk 12]
```diff
@@ -715,10 +576,8 @@ def _tensor_quant(inputs, amax, num_bits=8, unsigned=False, narrow_range=True):
 
     # Computation can be done in FP32 to prevent potential over flow.
     input_dtype = inputs.dtype
-    if inputs.dtype == torch.half:
-        inputs = inputs.float()
-    if amax.dtype == torch.half:
-        amax = amax.float()
+    inputs = inputs.float()
+    amax = amax.float()
```
[the remark this answers is not in the document]
Comment (realAsma) [reply modelopt/torch/quantization/tensor_quant.py:584]: done.

**Review by mxinO — APPROVED** [review]
LGTM! Thanks. Have you benchmarked the tirton vs CUDA on hopper+? we can use triton by default if it's faster.

File: modelopt/torch/quantization/tensor_quant.py [hunk 13]
```diff
@@ -42,10 +42,34 @@
 DISABLE_TRITON_KERNEL = False
 
 
+def _fp8_eager(x, amax=None):
```
Comment (cjluo-nv) [inline modelopt/torch/quantization/tensor_quant.py:45]: what would happen is amax is None?

Comment (meenchen) [reply modelopt/torch/quantization/tensor_quant.py:45]: What's the use case for None amax?

File: modelopt/torch/quantization/tensor_quant.py [hunk 14]
```diff
@@ -56,44 +80,22 @@ def scaled_e4m3_impl(
...
-            # and won't go back to normal value after dividing by scale. The right behavior is to mark them
-            # as zero which also get rid of inf/nan
-            outputs[zero_mask] = 0.0
-
-        return outputs
+        if amax.numel() == 1:
+            outputs = cuda_ext_fp8.fake_e4m3fy(inputs, amax)
+        elif amax.squeeze().ndim == 1:
+            axis = amax.shape.index(amax.numel())
+            outputs = cuda_ext_fp8.fake_e4m3fy_with_axis(inputs, amax.squeeze(), axis)
```
Comment (meenchen) [inline modelopt/torch/quantization/tensor_quant.py:89]: Should we add an else here to raise an error for ndim > 1?

File: modelopt/torch/quantization/tensor_quant.py [hunk 15]
```diff
@@ -56,44 +80,22 @@ def scaled_e4m3_impl(
...
-    def is_fusable():
-        # ignore no scaling and shape([]) cases
-        if amax is None or len(amax.shape) == 0:
-            return False
-        else:
-            # can't have amax.shape = [1, 1, 4, 1] and the like
-            amax_last_dim_only = amax.numel() == amax.shape[-1]
-            # must be cuda
-            all_cuda = inputs.is_cuda and amax.is_cuda
+    if (not inputs.is_cuda) or amax is None or amax.squeeze().ndim > 1:
```
Comment (realAsma) [inline modelopt/torch/quantization/tensor_quant.py:75]: > Should we add an else here to raise an error for ndim > 1?
@meenchen this is handled here.

File: modelopt/torch/quantization/tensor_quant.py (hunk 14 above)
Comment (realAsma) [reply modelopt/torch/quantization/tensor_quant.py:89]: see https://github.com/NVIDIA/TensorRT-Model-Optimizer/pull/381/files/ce384be73c995f94cc9eb8808eac503933b8f085#r2388899266

File: modelopt/torch/quantization/tensor_quant.py (hunk 13 above)
Comment (realAsma) [reply modelopt/torch/quantization/tensor_quant.py:45]: > What's the use case for None amax? 

Dynamic quantization without scaling.

Comment (realAsma) [reply modelopt/torch/quantization/tensor_quant.py:45]: > what would happen is amax is None?

QDQ without scaling will be performed

**Comment by realAsma** [comment]
@mxinO I am running into issues with torch.compile of FP8 eager mode. Let us implement torch.compile + FP8 eager mode in a seperate PR. Let me merge the current PR to fix the FP8 QAT bug on the main branch.
